File: Questions/arrays/rotate_image.py
# when rotating every thing in same idx gets grouped in 1 array
    # starts from last sub array to first sub array (right -> left)

#-------------------------------------plan-----------------------------
# create hash table
# create a key value pair from 0 to len of array - 1 
    # idx as a key, blank array as value
# loop through main array in reverse
# loop through sub array 
    # grab cur num 
    # append cur num to array at matching idx in hash table
# loop through main array replace each array with corresponding hash table array

def rotate(matrix):
    """
    Do not return anything, modify matrix in-place instead.
    """
    
    subArrs = {}
    
    for i in range(0,len(matrix)):
        subArrs[i] = []

    for i in reversed(range(len(matrix))):
        # print(matrix[i])
        for j, num in enumerate(matrix[i]):
            subArrs[j].append(num)
    # print(subArrs)
    
    for i in range(len(matrix)):
        matrix[i] = subArrs[i]
    # print(matrix)

# Time Complexity: O(n^2) 
    # where n is length of matrix array

# Space Complexity: O(n) 

File: Questions/arrays/rotate_image.js
function rotateImage(a) {
    let subArrs = {}
    
    for(let i = 0; i < a.length; i++) {
        subArrs[i] =  [];
    }
    
    for(let i = a.length-1; i >= 0; i--) {        
        for(let j = 0; j < a[i].length; j++) {
            subArrs[j].push(a[i][j]);
        }
    }
    
    for(let i = 0; i < a.length; i++) {
        a[i] = subArrs[i];
    }
    return a
}